[Shrink-To-Fit] shrink-to-fit sizing should consider max-inline-size: min-content.

sammygill opened the pull request:
#### 93eee6b40662353c53e00e0403410071dcf86e01
<pre>
[Shrink-To-Fit] shrink-to-fit sizing should consider max-inline-size: min-content.
<a href="https://bugs.webkit.org/show_bug.cgi?id=280549">https://bugs.webkit.org/show_bug.cgi?id=280549</a>
<a href="https://rdar.apple.com/136520853">rdar://136520853</a>

Reviewed by Alan Baradlay.

Currently, when we compute the preferred logical widths of a renderer, we will consider
the min/max logical width properties if they are fixed values. We should, however, be
able to handle any other values for these properties, and in this patch, we extend support
for min-content of the max logical width.

To accomplish this, I shuffled around some of the logic that was trying to compute the used
min/max logical widths at the top of the method. This should not be a functional change since
these computations don&apos;t rely on the values we have so far computed for m_min/maxPreferredLogicalWidth.

In the logic for computing the used max logical width, we now consider max-width: min-content:

- If the m_min/maxPreferredLogicalWidths computed from the style&apos;s logical width value, then
we compute the intrinsic logical widths of the renderer. This appears to happen when the
computed logical width is a fixed and positive value in most cases. RenderBlock has some
additional constraints in order to determine this, but I factored out this logic into a
separate method.

- Otherwise, m_min/maxPreferredLogicalWidths should hold the intrinsic values, so we can just
use the already computed value in m_minPreferredLogicalWidth.

(Also, a quick change of argument names in computePreferredLogicalWidths in the declaration
to match the ones used in the definition.)

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/shrink-to-fit-sizing-max-width-min-content-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/shrink-to-fit-sizing-max-width-min-content.html: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::shouldComputePreferredLogicalWidthsFromStyle const):
(WebCore::RenderBox::computePreferredLogicalWidths):
* Source/WebCore/rendering/RenderBox.h:

Canonical link: <a href="https://commits.webkit.org/286051@main">https://commits.webkit.org/286051@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/369aaee225036665ef0e70b72c98b78c4413c273

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74626 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54055 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27437 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79041 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25864 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76743 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63188 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1840 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58626 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16915 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77693 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48770 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64139 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39021 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45852 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21635 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24197 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67184 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21981 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80537 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1943 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1138 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66887 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2091 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64157 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66174 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16441 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10119 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8271 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1908 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4695 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1936 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/2857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1943 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3569,12 +3569,19 @@ void RenderBox::computePreferredLogicalWidths(const Length& minLogicalWidth, con
+        if (maxLogicalWidth.isFixed())
+            return adjustContentBoxLogicalWidthForBoxSizing(maxLogicalWidth);
+        if (maxLogicalWidth.isMinContent())
+            return computeIntrinsicLogicalWidthUsing(maxLogicalWidth, availableLogicalWidth(), 0_lu);
```
this looks a bit wasteful to me as in most cases we end up running preferred width computation here while we already have that value in m_minPreferredLogicalWidth -unless I am missing something.

sammygill replied:
That's a good point. I made a couple of changes:

1.) Reshuffled the logic to compute the used min/max logical widths to the top of the function to avoid m_minPreferredLogicalWIdth getting changed by the aspect ratio logic
2.) Refactored out some of the logic used to determine when preferred widths values will get set from the value in style so we could use it here

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3569,12 +3569,19 @@ void RenderBox::computePreferredLogicalWidths(const Length& minLogicalWidth, con
+            return adjustContentBoxLogicalWidthForBoxSizing(maxLogicalWidth);
+        if (maxLogicalWidth.isMinContent())
+            return computeIntrinsicLogicalWidthUsing(maxLogicalWidth, availableLogicalWidth(), 0_lu);
+        return LayoutUnit::max();
```
please avoid magic values if possible.

sammygill replied:
Changed!

alanbaradlay commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -713,7 +713,8 @@ class RenderBox : public RenderBoxModelObject {
     }
 
-    void computePreferredLogicalWidths(const Length& logicalMinWidth, const Length& logicalMaxWidth, LayoutUnit borderAndPaddingLogicalWidth);
+    bool shouldComputePreferredLogicalWidthsFromStyle() const;
```
make it private?

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3561,6 +3571,29 @@ void RenderBox::computePreferredLogicalWidths()
+                return m_minPreferredLogicalWidth;
+
+            static constexpr LayoutUnit s_noBorderAndPaddingLogicalWidth;
+            return computeIntrinsicLogicalWidthUsing(maxLogicalWidth, availableLogicalWidth(), s_noBorderAndPaddingLogicalWidth);
```
I think `computeIntrinsicLogicalWidthUsing(maxLogicalWidth, availableLogicalWidth(), { });` is a more readable way to express passing in default values even if it has some runtime overhead over the static variable which I don't think really matters here. -or assign a default value at the function declaration.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3551,6 +3551,16 @@ static bool allowMinMaxPercentagesInAutoHeightBlocksQuirk()
 
+bool RenderBox::shouldComputePreferredLogicalWidthsFromStyle() const
+{
+    if (isRenderBlock()) {
```
what makes "should compute from style" renderer type specific?

sammygill replied:
After taking another look I think maybe it should not be... I can' quite tell what's going on with the `!isRenderTableCell()` bit in `RenderBlock` though...

sammygill replied:
Actually yeah it probably doesn't make sense for `shouldComputePreferredLogicalWidthsFromStyle` either so I removed that specific renderer check too

sammygill committed:
[Shrink-To-Fit] shrink-to-fit sizing should consider max-inline-size: min-content.
https://bugs.webkit.org/show_bug.cgi?id=280549
rdar://136520853

Reviewed by Alan Baradlay.

Currently, when we compute the preferred logical widths of a renderer, we will consider
the min/max logical width properties if they are fixed values. We should, however, be
able to handle any other values for these properties, and in this patch, we extend support
for min-content of the max logical width.

To accomplish this, I shuffled around some of the logic that was trying to compute the used
min/max logical widths at the top of the method. This should not be a functional change since
these computations don't rely on the values we have so far computed for m_min/maxPreferredLogicalWidth.

In the logic for computing the used max logical width, we now consider max-width: min-content:

- If the m_min/maxPreferredLogicalWidths computed from the style's logical width value, then
we compute the intrinsic logical widths of the renderer. This appears to happen when the
computed logical width is a fixed and positive value in most cases. RenderBlock has some
additional constraints in order to determine this, but I factored out this logic into a
separate method.

- Otherwise, m_min/maxPreferredLogicalWidths should hold the intrinsic values, so we can just
use the already computed value in m_minPreferredLogicalWidth.

(Also, a quick change of argument names in computePreferredLogicalWidths in the declaration
to match the ones used in the definition.)

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/shrink-to-fit-sizing-max-width-min-content-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/shrink-to-fit-sizing-max-width-min-content.html: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::shouldComputePreferredLogicalWidthsFromStyle const):
(WebCore::RenderBox::computePreferredLogicalWidths):
* Source/WebCore/rendering/RenderBox.h:

Canonical link: https://commits.webkit.org/286051@main
